File: frontend/src/lib/permissions.ts
import type { PoFieldSpec, Role, TrackerColumn } from "./types";

/** Which pages and actions a role gets. Page-level only.
 *
 * Per-*field* editability is deliberately not here: the API decides it and
 * sends `editable` on every column and field spec (see `lockReason`). The
 * backend refuses the write either way, so there is nothing left for the UI
 * to work out for itself.
 *
 * What remains below mirrors app/permissions.py, and every route it guards is
 * enforced server-side by a dependency — a role that slips past one of these
 * gets a 403 from the endpoint.
 */

export const ROLE_LABELS: Record<Role, string> = {
  admin: "Admin",
  ceo: "CEO",
  shipping_manager: "Shipping Manager",
  merchant: "Merchant",
  vendor: "Vendor",
  pending: "Awaiting approval",
};

/** Roles an admin may assign in the user-management panel (not `pending`, which
 * is the absence of a role, nor the legacy `vendor`). Mirrors ALL_ROLES on the
 * backend. */
export const ASSIGNABLE_ROLES: { value: Role; label: string }[] = [
  { value: "merchant", label: "Merchant" },
  { value: "shipping_manager", label: "Shipping Manager" },
  { value: "ceo", label: "CEO" },
  { value: "admin", label: "Admin" },
];

/** Approve waitlisted accounts and assign roles - the admin panel. */
export const canManageUsers = (r: Role) => r === "admin";

/** The Shipment Tracker page itself, and every write to it. */
export const canViewTracker = (r: Role) =>
  r === "admin" || r === "ceo" || r === "shipping_manager";

/** Read-only tracker rows, for the dashboard's Shipment Tracker panel.
 *
 * Wider than `canViewTracker` on purpose: the panel's columns are all order
 * facts a merchant already sees per-PO. The tracker page, its export and its
 * edits are the shipping team's, and stay on `canViewTracker`.
 */
export const canReadTrackerRows = (r: Role) => canViewTracker(r) || r === "merchant";

/** The per-PO view: everyone who works orders, merchants included. */
export const canViewPos = (r: Role) =>
  r === "admin" || r === "ceo" || r === "shipping_manager" || r === "merchant";

export const canEditTracker = (r: Role) =>
  r === "admin" || r === "ceo" || r === "shipping_manager";

/** Buyer PO# / Style / Colour — changing these re-keys the row. Also the gate
 * on the Manual PO Line page, which creates a row and so sets its identity. */
export const canEditIdentity = (r: Role) => r === "admin" || r === "ceo";

export const canPaste = (r: Role) => r === "admin" || r === "shipping_manager";

export const canAssignSeason = (r: Role) =>
  r === "admin" || r === "ceo" || r === "merchant";

export const canUpload = (r: Role) =>
  r === "admin" || r === "merchant" || r === "ceo";

export const canViewAudit = (r: Role) => r === "admin" || r === "ceo";

export const canViewReports = (r: Role) => r === "admin" || r === "ceo";

export const canViewCustomers = (r: Role) => r === "admin" || r === "ceo";

export const canManageCustomers = (r: Role) => r === "admin";

export const canViewVendors = (r: Role) => r === "admin" || r === "ceo";

export const canManageVendors = (r: Role) => r === "admin";

export const canDelete = (r: Role) => r === "admin";

/** The spreadsheet-style grid, on the tracker page only. */
export const canUseExcelView = (r: Role) =>
  r === "admin" || r === "shipping_manager";

type Gated = Pick<
  TrackerColumn | PoFieldSpec,
  "editable" | "is_price" | "is_identity" | "is_payment_terms" | "is_derived"
> & { derived_from?: string[] };

/** Why a field is locked, for the tooltip on a disabled input — `null` when it
 * is not locked at all.
 *
 * `editable` settles *whether*; the flags only explain *why*, so a rule change
 * in permissions.py needs nothing here. The fallback covers a field locked for
 * a reason this UI has no wording for yet: still locked, just less helpfully.
 */
export const lockReason = (field: Gated): string | null => {
  if (field.editable) return null;
  if (field.is_derived) {
    // name the actual inputs: "change the price" is unhelpful advice on a
    // shipment-delay field. Two inputs (a subtraction or a total), or one (a
    // date offset). Wording stays operator-neutral so it fits all of them.
    const from = field.derived_from ?? [];
    if (from.length === 2)
      return `Calculated from ${from[0]} and ${from[1]}. Change those instead.`;
    if (from.length === 1)
      return `Calculated from ${from[0]}. Change that instead.`;
    return "Calculated automatically from other columns";
  }
  if (field.is_price)
    return "Price fields can only be changed by the CEO or an admin";
  if (field.is_identity)
    return "PO, style and colour identify the row — CEO or admin only";
  if (field.is_payment_terms)
    return "Payment terms are commercial — merchant, CEO or admin only";
  return "Read-only for your role";
};

const ORDER_DETAIL_SOURCES = ["buyer", "vendor", "const", "calc"];

export const isOrderDetailSource = (source: string) =>
  ORDER_DETAIL_SOURCES.includes(source);

File: backend/app/permissions.py
"""Role capabilities for the four user types.

Roles
-----
- ``admin``            - full access (edit everything, upload, delete, view audit).
- ``ceo``              - edits everything including prices, sees the audit trail,
                         may upload order sheets.
- ``shipping_manager`` - owns the Shipment Tracker and edits it freely, except
                         the price columns, the row's PO/Style/Colour identity,
                         and the two payment-terms columns.
- ``merchant``         - uploads order sheets, assigns each PO its season, and
                         works purchase orders in the per-PO view (same price and
                         identity limits). No Shipment Tracker: it is the shipping
                         team's table and irrelevant to them.

``vendor`` is a legacy role kept as a read-only viewer (orders only).

The editing model is deliberately flat: everyone who can edit, edits everything,
*except* the money columns, which are the CEO's and admin's alone, and the
payment terms, which belong to whoever negotiates them. Uploads are
not affected - a sheet that carries prices still writes them for whoever uploads
it, because that is the sheet's own figure and not a hand edit.
"""

from __future__ import annotations

from app.services import field_groups as fg
from app.services import tracker_map as tm

ADMIN = "admin"
CEO = "ceo"
SHIPPING_MANAGER = "shipping_manager"
MERCHANT = "merchant"

# A new account has no role until an admin assigns one: it signs up, lands on
# the waitlist, and passes none of the can_* checks below until approved.
PENDING = "pending"

# The roles an admin may assign. `pending` is deliberately absent - it is the
# absence of a role, not something you grant.
ALL_ROLES = (ADMIN, CEO, SHIPPING_MANAGER, MERCHANT)


def can_manage_users(role: str) -> bool:
    """Approve waitlisted accounts and assign roles - the admin panel."""
    return role == ADMIN


def can_view_tracker(role: str) -> bool:
    """The Shipment Tracker page itself, and every write to it."""
    return role in (ADMIN, CEO, SHIPPING_MANAGER)


def can_read_tracker_rows(role: str) -> bool:
    """Read-only tracker rows, for the dashboard's Shipment Tracker panel.

    Deliberately wider than ``can_view_tracker``: merchants see the panel, but
    the tracker page, the export and every tracker write stay shut to them,
    because those all hang off ``can_view_tracker``.
    """
    return can_view_tracker(role) or role == MERCHANT


def can_view_pos(role: str) -> bool:
    """The per-PO view: everyone who works orders, merchants included."""
    return role in (ADMIN, CEO, SHIPPING_MANAGER, MERCHANT)


def can_edit_prices(role: str) -> bool:
    """Money columns: buyer/factory price, both totals, price difference."""
    return role in (ADMIN, CEO)


def can_create_tracker_row(role: str) -> bool:
    """Who may bring a brand-new tracker row into existence.

    Merchants are here but not in ``can_edit_identity``: they may *set* a row's
    PO/Style/Colour once, when they create it, and may never change it
    afterwards - re-keying an existing row breaks re-import matching.
    Shipping edits rows but does not originate them.
    """
    return role in (ADMIN, CEO, MERCHANT)


def can_edit_identity(role: str) -> bool:
    """Buyer PO# / Style / Colour - changing these re-keys the row."""
    return role in (ADMIN, CEO)


def can_edit_payment_terms(role: str) -> bool:
    """Factory / Buyer Payment Terms - commercial, so not the shipping desk's."""
    return role in (ADMIN, CEO, MERCHANT)


def can_edit_tracker(role: str) -> bool:
    return role in (ADMIN, CEO, SHIPPING_MANAGER)


def can_paste(role: str) -> bool:
    """Pasting a table of PO details straight from an e-mail."""
    return role in (ADMIN, SHIPPING_MANAGER)


def can_assign_season(role: str) -> bool:
    return role in (ADMIN, CEO, MERCHANT)


def can_upload(role: str) -> bool:
    return role in (ADMIN, MERCHANT, CEO)


def can_view_audit(role: str) -> bool:
    return role in (ADMIN, CEO)


def can_view_reports(role: str) -> bool:
    return role in (ADMIN, CEO)


def can_view_customers(role: str) -> bool:
    return role in (ADMIN, CEO)


def can_manage_customers(role: str) -> bool:
    return role == ADMIN


def can_view_vendors(role: str) -> bool:
    return role in (ADMIN, CEO)


def can_manage_vendors(role: str) -> bool:
    return role == ADMIN


def can_delete(role: str) -> bool:
    return role == ADMIN


def editable_tracker_keys(role: str) -> frozenset[str]:
    """The set of tracker column keys this role is allowed to write."""
    if role not in (ADMIN, CEO, SHIPPING_MANAGER, MERCHANT):
        return frozenset()
    # price_difference is computed from the two prices - nobody hand-edits it
    keys = set(tm.TRACKER_KEYS) - set(tm.DERIVED_KEYS)
    if not can_edit_prices(role):
        keys -= tm.PRICE_KEYS
    if not can_edit_identity(role):
        keys -= tm.IDENTITY_KEYS
    if not can_edit_payment_terms(role):
        keys -= tm.PAYMENT_TERMS_KEYS
    return frozenset(keys)


def creatable_tracker_keys(role: str) -> frozenset[str]:
    """Editable keys plus identity, which is settable only at creation."""
    if not can_create_tracker_row(role):
        return frozenset()
    return frozenset(editable_tracker_keys(role) | tm.IDENTITY_KEYS)


def editable_line_keys(role: str) -> frozenset[str]:
    """Order-sheet line fields (article, garment spec...) this role may write."""
    if role not in (ADMIN, CEO, SHIPPING_MANAGER, MERCHANT):
        return frozenset()
    keys = set(fg.LINE_KEYS)
    if not can_edit_prices(role):
        keys -= fg.LINE_PRICE_KEYS
    return frozenset(keys)
